CI: Build AppImages

txtsd opened the pull request:
This PR makes it possible to start building AppImages on GHA.
The AppImages include the latest builds of jre8 and jre17, pulled directly from Adoptium via their [API](https://api.adoptium.net/q/swagger-ui/).
The AppImage building requires for a change in the .desktop file which is being discussed here #136.

It also changes the Windows and macOS builds to use Qt's `windeployqt` and `macdeployqt` tools respectively.

Additionally, it adds the commit SHA in the filenames.

EDIT: Adds a `Development Builds` section in `README.md`.
EDIT2: Now also builds 32bit PolyMC for Windows, instead of 64bit which fails to run for unknown reasons.
EDIT3: Also closes #136
EDIT4: Removes generic Linux builds because they don't necessarily work on all distros.

As always, please test the builds thoroughly before merging.

Scrumplex commented on .github/workflows/build.yml:
```diff
@@ -61,47 +68,102 @@ jobs:
+        run: |
+          DESTDIR=${{ env.INSTALL_DIR }} cmake --install ${{ env.BUILD_DIR }}
+
+      - name: Download linuxdeploy family
```
I think this should be moved up. We don't want to download dependencies after we compiled already. If this task fails we don't want to waste time building

txtsd replied:
Done!

Scrumplex commented on .github/workflows/build.yml:
```diff
@@ -61,47 +68,102 @@ jobs:
+        run: |
+          ${{ github.workspace }}/.github/scripts/prepare_JREs.sh
+
+      - name: AppImage magic
```
Maybe let's call this `Bundle AppImage` :D

txtsd replied:
Done!

Scrumplex commented on .github/workflows/build.yml:
```diff
@@ -61,47 +68,102 @@ jobs:
+          wget "https://github.com/linuxdeploy/linuxdeploy-plugin-appimage/releases/download/continuous/linuxdeploy-plugin-appimage-x86_64.AppImage"
+          wget "https://github.com/linuxdeploy/linuxdeploy-plugin-qt/releases/download/continuous/linuxdeploy-plugin-qt-x86_64.AppImage"
+
+      - name: Download JREs for AppImage
```
Move this up as well

txtsd replied:
Done!

Scrumplex commented:
LGTM!

binex-dsk commented on .github/workflows/build.yml:
```diff
@@ -60,48 +67,103 @@ jobs:
+          mkdir -p ${{ env.INSTALL_DIR }}/usr/lib/jvm/java-{8,17}-openjdk
+          cp -r ${{ github.workspace }}/JREs/jre8/* ${{ env.INSTALL_DIR }}/usr/lib/jvm/java-8-openjdk
+          cp -r ${{ github.workspace }}/JREs/jre17/* ${{ env.INSTALL_DIR }}/usr/lib/jvm/java-17-openjdk
+          LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:${{ env.INSTALL_DIR }}/usr/lib:${{ env.INSTALL_DIR }}/usr/lib/jvm/java-8-openjdk/jre/lib/amd64/server:${{ env.INSTALL_DIR }}/usr/lib/jvm/java-8-openjdk/jre/lib/amd64:${{ env.INSTALL_DIR }}/usr/lib/jvm/java-17-openjdk/lib/server:${{ env.INSTALL_DIR }}/usr/lib/jvm/java-17-openjdk/lib" \
```
this line hurts me, separate each entry by newlines

txtsd replied:
It's copying from folders that I create using the script in the `Download JREs for AppImage` step.

binex-dsk commented on .github/workflows/build.yml:
```diff
@@ -60,48 +67,103 @@ jobs:
+        shell: bash
         run: |
-          cmake --install build
+          export OUTPUT="PolyMC-${{ github.sha }}-x86_64.AppImage"
```
you can just do `export VERSION={{ github.sha }}` for the same effect

txtsd replied:
It's this way to conform with the other filenames, which are also named with `${{ github.sha }}`.

binex-dsk replied:
This would have that same effect. `VERSION=${{ github.sha }}` can be used instead and it's shorter so I like it.

binex-dsk commented on .github/workflows/build.yml:
```diff
@@ -60,48 +67,103 @@ jobs:
-          cmake --install build
+          export OUTPUT="PolyMC-${{ github.sha }}-x86_64.AppImage"
+          chmod +x linuxdeploy-*.AppImage
+          mkdir -p ${{ env.INSTALL_DIR }}/usr/lib/jvm/java-{8,17}-openjdk
```
It feels really weird to put JREs in a JDK directory, are you copying a `jre` folder?

This also reminds me that the LD_LIBRARY_PATH might be wrong because of that. What is the directory structure of the Adoptium JREs?

txtsd replied:
No, the JREs are downloaded directly from Adoptium. Their structures are the same as their JDK alternatives. JRE8 has `bin`, `lib`, `man`, and JRE17 has `bin`, `lib`, `conf` and `legal`.
I have no JRE folders in my `/usr/lib/jvm` even though I have JREs installed, so I put them in the JDK folders like they are on my system.

txtsd replied:
Oh you might be right about the `LD_LIBRARY_PATH`. Let me change that.

txtsd replied:
Only the jre8 path was incorrect in `LD_LIBRARY_PATH`. Fixed it and split up the lines too.

txtsd commented:
Could you double-check the installed JREs in the AppImage? It includes 17.0.2, but I only see 17.0.1 from my system.

binex-dsk reviewed (changes requested):
small at a glance review to prevent myself from getting eye cancer, and the weird JRE dir structure

binex-dsk commented on .github/scripts/prepare_JREs.sh:
```diff
@@ -0,0 +1,49 @@
+do
+    mkdir temp
+
+    # Handle OpenJDK17 archive
```
This whole thing could probably be condensed into a for loop (`for n in 8 17`). The two lines that are different can be put in an if statement

txtsd replied:
No, it needs to read the file names to handle regex for the version numbers. We're pulling latest builds, not a pinned version. I de-duped the code, regardless.

binex-dsk commented on .github/workflows/build.yml:
```diff
@@ -60,48 +67,113 @@ jobs:
+          LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:${{ env.INSTALL_DIR }}/usr/lib/jvm/java-17-openjdk/lib/server"
+          LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:${{ env.INSTALL_DIR }}/usr/lib/jvm/java-17-openjdk/lib"
+
+          ./linuxdeploy-x86_64.AppImage --appdir ${{ env.INSTALL_DIR }} --output appimage --plugin qt -d ${{ env.INSTALL_DIR }}/usr/share/applications/org.polymc.polymc.desktop -i ${{ env.INSTALL_DIR }}/usr/share/icons/hicolor/scalable/apps/org.polymc.PolyMC.svg
```
I don't think you technically need the `-d`. It can be kept if a more explicit specification is desirable

txtsd replied:
Got rid of the `-d` since it works the same.

binex-dsk commented on .github/workflows/build.yml:
```diff
@@ -60,48 +67,113 @@ jobs:
+        if: matrix.app_image == true
+        uses: actions/upload-artifact@v2
+        with:
+          name: PolyMC-${{ github.sha }}-x86_64.AppImage
```
why does it need the whole thing twice, is it just a github moment

txtsd replied:
Yea, it names itself `artifact` if no name is provided.

binex-dsk reviewed (commented):
more nitpicky eye candy stuff

binex-dsk commented on .github/scripts/prepare_JREs.sh:
```diff
@@ -0,0 +1,43 @@
+        fi
+        echo $hyphen
+
+        version_edit=$(echo $version_trailing | sed -e 's/_/+/g' | sed -e 's/b/-b/g')
```
I'm not a big fan of piping from echo. You can do something like:

```
sed -e ... <<< `sed -e <<< ...`

```

To be fair that's still kinda cancer. In the end, it's up to personal preference, and the current code looks perfectly fine.

txtsd replied:
Me either, but the pipes will be friendlier to any other dev who might work on it.

binex-dsk commented on .github/scripts/prepare_JREs.sh:
```diff
@@ -0,0 +1,43 @@
+        else
+            hyphen=''
+        fi
+        echo $hyphen
```
left in from debugging?

txtsd replied:
Yea, whoops. Got rid of the `echo`s.

binex-dsk reviewed (approved):
Remove that extraneous echo and this LGTM to me.

Don't merge yet though, I'll have to test all builds to ensure they work. Though I'm not sure if my 2012 Macbook supports a new enough version (is Catalina new enough to use this?)

ghost commented:
One more thing: it should be mentioned on the README that per-commit builds are available (as "dev" builds). This is especially needed for MacOS as we now have (theoretically) functional packages for it.

Maybe in the future we can implement {Free,Open}BSD builds but that's for a future PR.

txtsd commented:
Okay let me add to the README and then we'll be good to go.

The macOS builds target `10.12` so you should be good.

txtsd commented:
Okay README changes are in!

txtsd commented:
Final commit for building 32bit PolyMC on windows instead of 64bit.

txtsd commented:
Hold up on this. Checking to see why the Linux build is failing with:
```
λ ./PolyMC
Launcher Dir: /home/txtsd/Downloads/polymc-linux
No missing dependencies found.
/home/txtsd/Downloads/polymc-linux/polymc: symbol lookup error: /home/txtsd/Downloads/polymc-linux/polymc: undefined symbol: _ZdlPvm, version Qt_5
```

txtsd commented:
Okay I removed generic Linux builds because they don't necessarily work on all distros.

The OpenSSL step is now split into two. It downloads before building, and copies later.

Also added missing flags to `windeployqt` to not include translations, and to `macdeployqt` to include debug versions of frameworks and plugins.

ghost commented:
> Okay I removed generic Linux builds because they don't necessarily work on all distros.

why, if it's lin-nodeps it should work perfectly fine

txtsd commented:
@binex-dsk It was failing for me as well as others with the error posted above. Something about mismatched Qt libraries.

dada513 reviewed (approved):
LGTM!

txtsd commented:
Had to rebase on develop.

d-513 commented:
@txtsd java detection is broken in the appimages

txtsd commented:
@dada513 What can't it find? Included java or system java?

d-513 commented:
> @dada513 What can't it find? Included java or system java?

both, it cant find any javas

txtsd commented:
Might be a distro specific thing. See #117 
Can you test on any other distro?

d-513 commented:
> Might be a distro specific thing. See #117 Can you test on any other distro?

currently on ubunut 21.10 but am switching distros anyway so I will test soon
